src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -603,6 +611,155 @@ fromChunks cs FinalChunk{..} = CompactIndex {
     , ciLargerThanPage       = fcLargerThanPage
     }
 
+{-------------------------------------------------------------------------------
+  Serialisation
+-------------------------------------------------------------------------------}
+
+-- | 32 bit aligned.
+chunkBuilder :: Chunk -> BB.Builder
+chunkBuilder Chunk {..} = VU.foldMap BB.word32LE cPrimary
+
+-- | Must be written after the sequence of 'chunkBuilder' of the corresponding
+-- compact index. Specifically, the written chunks must match 'fcNumPages' to
+-- get the alignment right.
+finalChunkBuilder :: NumEntries -> FinalChunk -> BB.Builder
+finalChunkBuilder (NumEntries numEntries) FinalChunk {..} =
+       VU.foldMap BB.word32LE fcRangeFinder
+    <> (if odd (fcNumPages + VU.length fcRangeFinder)  -- align to 64 bit
+        then BB.word32LE 0
+        else mempty)
+    <> putBitVec fcClashes
+    <> putBitVec fcLargerThanPage
+    <> putTieBreaker fcTieBreaker
+    <> BB.word64LE (fromIntegral fcRangeFinderPrecision)
+    <> BB.word64LE (fromIntegral fcNumPages)
+    <> BB.word64LE (fromIntegral numEntries)
+
+-- | Padded to 64 bit.
+--
+-- Assumes that the bitvector is byte-aligned.
+putBitVec :: VU.Vector Bit -> BB.Builder
+putBitVec (BitVec offsetBits lenBits ba)
+  | mod8 offsetBits /= 0 = error "putBitVec: not byte aligned"
+  | otherwise =
+       -- first only write the bytes that are fully part of the bit vec
+       byteArrayFromTo offsetBytes offsetLastByte ba
+       -- then carefully write the last byte, might be partially uninitialised
+    <> (if remainingBits /= 0 then
+         BB.word8 (indexByteArray ba offsetLastByte .&. bitmaskLastByte)
+       else
+         mempty)
+    <> putPaddingTo64 totalBytesWritten
```

Comment:
This works alright, but I am considering switching to a different bitvector type that is closer to what we want, just as @phadej did for the Bloom filter. We need a little more functionality than the `BitVec64` provides, but that should be quick to add.

Reply:
> It might not be byte aligned, which would now just fail at runtime. 

Byte-aligned in what sense? My assumption is that byte arrays are *at least* 1-byte aligned, but n-byte aligned if you specify it explicitly

> If BitVec64 for some reason is slower than bitvec, then we should run into this in the bloom filter already.

I'm not sure we comprehensively compared the performance of the bloom filter pre- and post-`BitVec64`-ification

Reply:
It's a bit vector and the offset is in number of bits. So you could do `drop 4` and now the first byte of the underlying byte array is half used by the bit vector and half unused. So when serialising, we'd need to consider that.

Reply:
Okay, so it's the offset that isn't necessarily byte-aligned and not the the underlying byte array itself. So maybe we should instead say:

```
-- Assumes that the bitvector has a byte-aligned offset.
```

Or alternatively, add a `BB.word8` for the first byte in case the offset isn't byte-aligned, and remove the guard

Reply:
I'll change the comment,

However, it's not as easy as handling the first byte in a smart way, since all the other bytes need to shift by some number of bits as well. Each byte of output would need to be produced by reading two bytes, shift and bitwise or.

Reply:
In any case, I think it's not necessary to add my suggestion -- I'm fine with it as is, I have no strong preference

Reply:
> I think it's fine to have some unused bits at the front

Yes, but then we'd also need to serialise the offset so we know that the bits are unused. It's all solvable, but I don't think it's something we need to deal with.